### DISCUSSION TITLE
Flip of interfaces not properly working after mirror

### DISCUSSION THREAD

**Discussion by Joostth98** [mjhoptics/ray-optics#131] [Q&A]
After placing a mirror and rotating it as follows. 

`opm.add_mirror(label='1',t = -100)`
`sm.ifcs[-2].decenter = srf.DecenterData('bend', alpha=-45)`

And then adding a relay lens (triplet example from the docs) 

`opm.add_from_file('Parts/RelayLens.roa', t = -50)`

 We get the following results
![image](https://github.com/mjhoptics/ray-optics/assets/76204849/7ecb2e6c-9ea6-40a3-ba4b-734c64d404bb)

As we can see this is not the desired result.
If we now flip the relay system, we get .
`opm.flip(6,11)`

![image](https://github.com/mjhoptics/ray-optics/assets/76204849/007bd108-0f6c-48dd-a5cc-867d5a6d2e8a)
And still we do not have the desired result. How can I fix this?

**Comment by mjhoptics** [comment]
# Investigate issue #131

The issue is described as a fold mirror followed by a "inserted" optical system doesn't give the expected results. 

This notebook shows a set of steps that will produce a model that gives the expected results. To be fair, these steps are non-intuitive at best (NIAB)

The Sasian Triplet from the rayoptics/models directory is used for the inserted optical system.


```python
%matplotlib inline
isdark = False
from rayoptics.environment import *
import itertools
root_pth = Path(rayoptics.__file__).resolve().parent
```


```python
opm = OpticalModel()
opm['system_spec'].title = 'Issue 131 example'
sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
```

Use the same optical specs as are used by the Sasian Triplet. 

The fold mirror in front of the lens needs to be sized by tracing the top and bottom field points.


```python
osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=12.5)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=20., flds=[0., -1., 1.], is_relative=True)
osp['wvls'] = WvlSpec([("F", 0.5), ('d', 1.0), ('C', 0.5)], ref_wl=2)
```

```python
opm.radius_mode = True
sm.do_apertures = False

sm.gaps[0].thi=1e10
```

A dummy surface is added first. This will be used as the reference coordinate system for lens pictures, global coordinates, etc.

I list the sequential model and do an update_model() after each change.


```python
sm.add_surface([0, 0, 'air', 25.])
sm.list_model()
opm.update_model(src_model=sm)
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:     0.000000      0.00000       air             1      25.000
      Img:     0.000000      0.00000                       1      1.0000


Add the mirror and define it as a Bend surface; update the model.


```python
opm.add_mirror(label='M1',t = -100, sd=100.)
sm.ifcs[-2].decenter = srf.DecenterData('bend', alpha=-45)

opm.update_model()
```

List the decenter data and the sequential model.

List the global coordinates and rotation matrix for each surface with respect to the first surface.


```python
sm.list_sg()
print()
sm.list_gbl_tfrms()
```

                   r               mode              type          y       alpha
                           t           medium
      Obj:      0.00000                 
                     1.00000e+10          air
        1:      0.00000                 
                         0.00000          air
        2:      0.00000          reflect               bend     0.0000    -45.000
                        -100.000          air
      Img:     -0.00000                 
    
     0:    1.000000    0.000000    0.000000       0.00000
           0.000000    1.000000    0.000000       0.00000
           0.000000    0.000000    1.000000  -10000000000.00000
    
     1:    1.000000    0.000000    0.000000       0.00000
           0.000000    1.000000    0.000000       0.00000
           0.000000    0.000000    1.000000       0.00000
    
     2:    1.000000    0.000000    0.000000       0.00000
           0.000000    0.707107   -0.707107       0.00000
           0.000000    0.707107    0.707107       0.00000
    
     3:    1.000000    0.000000    0.000000       0.00000
           0.000000    0.000000   -1.000000     100.00000
           0.000000    1.000000    0.000000      -0.00000
    


Add the triplet following the fold mirror. The thickness following the triplet is entered as a positive value; we'll fix it up later.


```python
opm.add_from_file(root_pth/'models/Sasian Triplet.roa', t = 50)
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:     0.000000      0.00000       air             1      25.000
        2:     0.000000     -100.000       air    reflect -1      100.00
        3:    23.713000      4.83100    N-LAK9            -1      10.009
        4:  7331.288000      5.86000       air            -1      8.9482
        5:   -24.456000     0.975000     N-SF5            -1      4.7919
        6:    21.896000      4.82200       air            -1      4.7358
        7:    86.759000      3.12700    N-LAK9            -1      8.0217
        8:   -20.494200      50.0000       air            -1      8.3321
      Img:    -0.000000      0.00000                      -1      1.0000


The aperture stop is at the central, negative element of the triplet. Set the stop surface at that element.


```python
sm.set_cur_surface(5)
sm.set_stop()
```


The sign convention for surfaces following a reflection is that the thicknesses are negative. The sm.list_model() above shows the airspace following the mirror is -100 but the triplet spacings are positive. 

Starting at the first lens surface, scale the remaining surfaces by -1 and update the model.


```python
start = 3

seq = itertools.zip_longest(sm.ifcs[start:], sm.gaps[start:])

for i, sg in enumerate(seq, start=start):
    ifc, g = sg
    ifc.apply_scale_factor(-1)
    if g:
        g.apply_scale_factor(-1)
```


```python
opm.update_model()
```


```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:     0.000000      0.00000       air             1      25.000
        2:     0.000000     -100.000       air    reflect -1      100.00
        3:   -23.713000     -4.83100    N-LAK9            -1      10.009
        4: -7331.288000     -5.86000       air            -1      8.9482
     Stop:    24.456000    -0.975000     N-SF5            -1      4.7919
        6:   -21.896000     -4.82200       air            -1      4.7358
        7:   -86.759000     -3.12700    N-LAK9            -1      8.0217
        8:    20.494200     -50.0000       air            -1      8.3321
      Img:     0.000000      0.00000                      -1      1.0000



```python
listobj(osp)
```

    aperture: object epd; value=12.5
    field: object angle; value=20.0
    x,y=0.0 vlx= 0.000 vux= 0.000 vly= 0.000 vuy= 0.000
    y=-1.0 vly= 0.000 vuy= 0.000 vlx= 0.000 vux= 0.000
    y=1.0 vly= 0.000 vuy= 0.000 vlx= 0.000 vux= 0.000
    central wavelength=  656.2725 nm
    wavelength (weight) =  486.1327 (0.500),  587.5618 (1.000),  656.2725 (0.500)*
    focus shift=0.0
    



```python
pm.first_order_data()
```

    efl              -50.05
    f                 50.05
    f'               -50.05
    ffl               62.83
    pp1               112.9
    bfl              -41.29
    ppk                8.76
    pp sep           -223.7
    f/#               4.004
    m            -5.005e-09
    red          -1.998e+08
    obj_dist          1e+10
    obj_ang              20
    enp_dist          111.7
    enp_radius         6.25
    na obj         6.25e-10
    n obj                 1
    img_dist         -41.29
    img_ht            18.22
    exp_dist          18.71
    exp_radius        6.405
    na img           0.1239
    n img                -1
    optical invariant        2.275


Change the backfocus to be the paraxial bfl


```python
sm.gaps[-1].thi = ar['parax_data'].fod.bfl
opm.update_model()
```


```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=isdark).plot()
```


    
![output_25_0](https://github.com/mjhoptics/ray-optics/assets/33294031/13195855-19d2-456d-929e-72da2182b84e)
    



```python
ta_plt = plt.figure(FigureClass=RayFanFigure, opt_model=opm, data_type='Ray', 
                     scale_type=Fit.User_Scale, user_scale_value=.05, dpi=130).plot()
```


    
![output_26_0](https://github.com/mjhoptics/ray-optics/assets/33294031/fc87edbc-f32c-45b0-b9dd-2c6e78096865)
    



```python
wav_plt = plt.figure(FigureClass=RayFanFigure, opt_model=opm, data_type='OPD', 
                     scale_type=Fit.All_Same, user_scale_value=0.5, dpi=130).plot()
```


    
![output_27_0](https://github.com/mjhoptics/ray-optics/assets/33294031/97b63f0a-67ae-4824-8105-cf3b911b7f77)
    



```python

```

**Comment by lkangas** [comment]
Hi Mike. 

I'm trying to apply the above instructions for an Edmund catalog lens, a plano convex even asphere.

I'm trying to fold the optical axis back on itself, so that I get refraction trough the same lens again.

In the non-tilted fold, my lens changes shape. I experimented with 45 and 30 degree folds (commented out in the following script), which produce even worse results, where it seems that maybe something is rendered wrong. In this example, however, the traced rays match the rendering, but the lenses don't match each other. 

Are the asphere coeffs not considered in the flipping and/or -1 scales?

![image](https://github.com/user-attachments/assets/5f9e3466-6b1a-41cc-8d94-d55db5860bb9)


```python

import matplotlib

matplotlib.use('Qt5Agg')
import matplotlib.pyplot as plt

plt.ion()

from rayoptics.environment import OpticalModel, FieldSpec, PupilSpec
from rayoptics.environment import InteractiveLayout


opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
em = opm['ele_model']
ar = opm['analysis_results']
pm = opm['parax_model']

osp['pupil'] = PupilSpec(osp, key=['object', 'epd'], value=7.)
osp['fov'] = FieldSpec(osp, key=('object', 'angle'), value=0.0, flds=[1.], is_relative=True)

sm.do_apertures = False

sm.gaps[0].thi = 1e10

opm.add_from_file('CODV_48176.seq', t=1)

bfl = ar['parax_data'].fod.bfl
sm.gaps[-1].thi = bfl

opm.add_mirror(label='M1', t=-bfl)
# sm.ifcs[-2].decenter = srf.DecenterData('bend', alpha=45)

opm.add_from_file('CODV_48176.seq', label='L2', t=20)
lens22 = em.elements[-3]

opm.flip(lens22)
opm.update_model()

start = 4
import itertools

seq = itertools.zip_longest(sm.ifcs[start:], sm.gaps[start:])

for i, sg in enumerate(seq, start=start):
    ifc, g = sg
    ifc.apply_scale_factor(-1)
    if g:
        g.apply_scale_factor(-1)

opm.update_model()

layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
                        do_draw_rays=False, do_paraxial_layout=False,
                        do_draw_beams=False, do_draw_ray_fans=True, num_rays_in_fan=7).plot()
```

**Reply by mjhoptics** [reply]
Hello @lkangas,
I've checked in a fix cb394e59b966a9aeb743d54586e3a0dfa6e3a367 in on the master branch. The problem was in the apply_scale_factor function for aspheric profiles. I also fixed an unrelated bug raytracing aspheric toroids.
I'll update a notebook with an example in the next day or so. We've been having an "event" here in Southern California and  it's ... incomprehensible.

**Reply by lkangas** [reply]
Thank you Mike. Let me test and I'll confirm the fix is working. I can also demonstrate whether the problems with bend angles persists, which I didn't show example figures for above.

I really wish you stay safe, and others around you as well.

**Reply by lkangas** [reply]
The above code now works with the latest master branch, great!

![image](https://github.com/user-attachments/assets/78124114-6ad0-4c50-9395-7efa7c951649)

After adding `sm.ifcs[-2].decenter = srf.DecenterData('bend', alpha=45)`, I also get what I expect, if I exclude the flipping. Works also for  `alpha=30`.

![image](https://github.com/user-attachments/assets/d32bf7e3-6685-45a1-a777-18a7e2c7d5ae)
![image](https://github.com/user-attachments/assets/a7ba8578-77d0-4918-beff-a1d51d51ed47)

This works for other bend angles too, e.g. `alpha=30`. However, flipping the lens produces weird stuff:

![image](https://github.com/user-attachments/assets/287d0960-bc0e-4418-b533-884955f72a92)
![image](https://github.com/user-attachments/assets/10263226-ef21-42ea-b505-191f9154401d)

Judging by the collimation of the beams, the ray tracing is correct, but maybe the part rendering isn't? Looks like they need a vertical flip around the horizontal line crossing the curved surface vertex.

Lauri

**Reply by mjhoptics** [reply]
Hello Lauri,
I have a checkin on the master branch that fixes the issue, e11e733a90244ebadbd8387cf2d76e95568e07ea. You were correct, the flip following a non-zero rotation was being applied incorrectly. Coordinate transformations are so picky! Thanks for documenting the issue clearly.
Mike

**Reply by lkangas** [reply]
Oh yes, I'm still waiting for the day a rotation matrix works in the correct direction on the first try..

I can confirm the fix is working. Thanks again! This library really helps both at work and with some astronomy projects in the spare time.

![image](https://github.com/user-attachments/assets/d5f19d25-1f5b-4a1e-b91e-c79f952f90a3)
